### ISSUE TITLE
Differential checks (downloads) a very big (not uploaded) folder in the root

### ISSUE THREAD

**Issue by dorongutman** [MathieuLoutre/grunt-aws-s3#69]
I have the following structure in the root of my bucket:

img/
css/
very-big-folder/
index.html

In my local folder the very-big-folder doesn't exist. However, it still looks like it is downloaded to be evaluated for differential matches.
I'm using version 0.9.4.

**Comment by MathieuLoutre** [comment]
This is strange. I need to take a deeper look at it, can't really see what it can be (besides a bug) from the top of my head. Can you try the latest version?

**Comment by dorongutman** [comment]
I've tried it now with the latest version, and the same thing happens.
Can you add an "exclude" list for differential checks maybe ?

**Comment by MathieuLoutre** [comment]
Hi, sorry for the slow reply. I haven't had the time to investigate, I still suspect it's a bug in the code that should be able to be fixed without adding features. However, I could be wrong, and in this case adding an `exclude` list would be a great solution.

**Comment by dorongutman** [comment]
This is what I see:
In the upload task, you list the contents of the bucket - https://github.com/MathieuLoutre/grunt-aws-s3/blob/0.14.0/tasks/aws_s3.js#L806 **recursively** (according to https://github.com/MathieuLoutre/grunt-aws-s3/blob/0.14.0/tasks/aws_s3.js#L329-L331)

**Comment by MathieuLoutre** [comment]
`listObjects` will indeed look at everything in the bucket that starts with the given prefix (which in the case of the differential upload is the `dest`).  In some cases `dest` can be the root and hence the whole bucket, but even so, `listObjects` isn't supposed to download anything, particularly when uploading, it's only listing the files on the bucket and uses MD5 hash returned to check what exists already.

Is your file downloaded on your hard drive or is it listed as part of the log?

**Comment by dorongutman** [comment]
So if I have a folder in my destination bucket (which is not in the input folder on my computer), and that folder has millions of files in it - it'll take a lot of time to go over all of these files, even though that folder doesn't interest me as I don't upload anything to it.

So we do must have an "exclude" list of what to ignore - even from just listing it for MD5 hashes.

**Comment by MathieuLoutre** [comment]
Right, makes sense now! I guess if you can't avoid using the root as a dest, you do need to be able to exclude some paths. However, I'm not sure how to best do that:
- I can only list files/folders using a prefix
- if there's not prefix it'll scan everything
- in your case your config doesn't seem to have a yield a list of prefixes
- When listing with no prefix, AWS doesn't provide a way to filter the results/exclude a path

A potential solution would be to list to the top files/folders and try to work out the exclude from there but it seems like a pretty crazy thing to do (what about depth? and if you have a million top folders?)

Maybe it would come down to be able to provide the folders to check with differential rather than the ones to not check? How does that sound?

**Comment by dorongutman** [comment]
If the user provides the source folder - why not list only the root (using the prefix), and then only the folders that are in the source folder ? 

So if my local source folder is something like this:

- FolderOne
- FolderTwo
- FileOne.ext
- FileTwo.ext

and my S3 bucket contains:

- myPrefix/FolderOne
- myPrefix/FolderTwo
- myPrefix/FolderThree
- myPrefix/FileOne.ext
- myPrefix/FileTwo.ext
- myPrefix/FileThree.ext

and the prefix is "myPrefix", you list all files (not recursively) in S3 under "myPrefix", and then you only list (recursively) myPrefix/FolderOne and myPrefix/FolderTwo, but not myPrefix/FolderThree as it's not in the source folder.

**Comment by dorongutman** [comment]
Ping

**Comment by MathieuLoutre** [comment]
Sorry I've neglected this issue, I haven't had much time to focus on this project lately.
Can you try to `console.log(unique_dests)` on L805 and L800 of the latest version? (https://github.com/MathieuLoutre/grunt-aws-s3/blob/0.14.1/tasks/aws_s3.js#L805, https://github.com/MathieuLoutre/grunt-aws-s3/blob/0.14.1/tasks/aws_s3.js#L800)

I'd like to see what the current filtering system isn't doing well. If there's nothing in the root folder that you're trying to upload, it shouldn't scan the whole thing.
